zerver/migrations/0699_fix_user_role_system_groups.py

```diff
@@ -0,0 +1,191 @@
+    UserGroupMembership = apps.get_model("zerver", "UserGroupMembership")
+
+    def get_system_user_group_by_name(group_name: str, realm_id: int) -> Any:
+        return NamedUserGroup.objects.get(name=group_name, realm_id=realm_id, is_system_group=True)
+
+    def realm_user_count_by_role(realm: Any) -> dict[str, Any]:
+        """
+        Code copied from the implementation in zerver.lib.user_counts
+        """
+        ROLE_COUNT_HUMANS = "11"
+        ROLE_COUNT_BOTS = "12"
+
+        human_counts = {
+            str(ROLE_REALM_ADMINISTRATOR): 0,
+            str(ROLE_REALM_OWNER): 0,
+            str(ROLE_MODERATOR): 0,
+            str(ROLE_MEMBER): 0,
+            str(ROLE_GUEST): 0,
+        }
+        for value_dict in (
+            UserProfile.objects.filter(realm=realm, is_bot=False, is_active=True)
+            .values("role")
+            .annotate(Count("role"))
+        ):
+            human_counts[str(value_dict["role"])] = value_dict["role__count"]
+        bot_count = UserProfile.objects.filter(realm=realm, is_bot=True, is_active=True).count()
+        return {
+            ROLE_COUNT_HUMANS: human_counts,
+            ROLE_COUNT_BOTS: bot_count,
+        }
+
+    def fix_user_memberships_if_needed(user_profile: Any) -> None:
+        realm = user_profile.realm
+
+        guest_group = get_system_user_group_by_name(SystemGroups.EVERYONE, realm.id)
+        member_group = get_system_user_group_by_name(SystemGroups.MEMBERS, realm.id)
+        moderator_group = get_system_user_group_by_name(SystemGroups.MODERATORS, realm.id)
+        admin_group = get_system_user_group_by_name(SystemGroups.ADMINISTRATORS, realm.id)
+        owner_group = get_system_user_group_by_name(SystemGroups.OWNERS, realm.id)
+
+        group_id_to_group_name = {
+            guest_group.id: "Guests",
+            member_group.id: "Members",
+            moderator_group.id: "Moderators",
+            admin_group.id: "Administrators",
+            owner_group.id: "Owners",
+        }
+        role_to_group = {
+            ROLE_GUEST: guest_group,
+            ROLE_MEMBER: member_group,
+            ROLE_MODERATOR: moderator_group,
+            ROLE_REALM_ADMINISTRATOR: admin_group,
+            ROLE_REALM_OWNER: owner_group,
+        }
+        group_id_to_role = {v.id: k for k, v in role_to_group.items()}
+
+        role_group_ids = [
+            guest_group.id,
+            member_group.id,
+            moderator_group.id,
+            admin_group.id,
+            owner_group.id,
+        ]
+
+        role = user_profile.role
+        group_implied_by_role = role_to_group[role]
+
+        direct_memberships = UserGroupMembership.objects.filter(
+            user_profile=user_profile, user_group_id__in=role_group_ids
+        )
+        if len(direct_memberships) == 0:
+            print(
+                f"User {user_profile.id} has no role group memberships. This is unexpected. Skipping."
```

Comment:
Can't we just add the user to the group as per their role instead of skipping?

Reply:
I think we can, but the idea was for this migration to fix the state resulting from the ldap bug. "User has no memberships" (or too many memberships) is not a state that we know to be possible due to that bug.

So I didn't want to mess with that and try to fix it silently, since if we encounter such a state, we aren't really super sure how it occurred and what the fix should be.

Would be an easy tweak here though, if we think we want it in this migration.

Reply:
Understood. Seems fine to just fix the incorrect state from ldap bug.